### PR TITLE
Update to support v5 data format

### PR DESCRIPTION
**Pull request by kevinlind**
## Description

Update to support new data format as defined below. This is a breaking change as the old data format is no longer supported. This feature is to be released in the next major release v5.0.0.

Changes how the track event data is mapped to an experience event. The new data format maps the data under the `data.__adobe.analytics` object. This change allows for the data to get automatically mapped to known Analytics fields.

 The following is the mapping logic:
  - The "action" field is mapped to "data.__adobe.analytics.linkName", plus "data.__adobe.analytics.linkType" is set to "other".
  - The "state" field is mapped to "data.__adobe.analytics.pageName".
  - Any "contextData" keys which use the "&&" prefix are mapped to "data.__adobe.analytics" with the prefix removed.
  - Any "contextData" keys which do not use the "&&" prefix are mapped to "data.__adobe.analytics.contextdata".
  - Any additional fields are passed through and left directly under the "data" object.
   
  As an example, the following track event data:
 ```json
      {
         "action": "action name",
         "contextdata": {
            "&&c1": "propValue1",
            "key1": "value1"
         }
         "key2": "value2"
      }
  ```
  Is mapped to:
  ```json
      {
        "data": {
          "__adobe": {
            "analytics": {
              "linkName": "action name",
              "linkType": "other",
              "c1": "propValue1"
              "contextData": {
                "key1": "value1"
             }
            }
          }
         "key2": "value2"
        }
      }
 ```

Null or empty string keys are not allowed. Values must be of type String, Number, or Character, though empty String values are allowed.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](https://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: Sources/EdgeBridge.swift [hunk 1]
```diff
@@ -107,4 +107,75 @@ public class EdgeBridge: NSObject, Extension {
...
+        var mutableData = data // mutable copy of data
+        var analyticsData: [String: Any] = [:] // __adobe.analytics data
+
+        if let contextData = mutableData.removeValue(forKey: "contextdata") as? [String: Any], !contextData.isEmpty {
+            let prefixedDataArray = contextData.filter { $0.key.hasPrefix("&&") }.map { (String($0.dropFirst("&&".count)), $1) }
+            analyticsData = Dictionary(uniqueKeysWithValues: prefixedDataArray) as? [String: Any] ?? [:]
+
+            let nonprefixedData = contextData.filter { !$0.key.hasPrefix("&&") }
+            if !nonprefixedData.isEmpty {
+                analyticsData["contextdata"] = nonprefixedData
```
Comment (emdobrin) [inline Sources/EdgeBridge.swift:161]: The destination key should be `contextData`, make it constant too.

File: Tests/UnitTests/EdgeBridgeTests.swift [hunk 2]
```diff
@@ -48,14 +48,248 @@ class EdgeBridgeTests: XCTestCase, AnyCodableAsserts {
         XCTAssertTrue(edgeBridge.readyForEvent(event))
     }
 
-    func testHandleTrackEvent_dispatchesEdgeRequestEvent() {
+    func testHandleTrackEvent_withActionField_dispatchesEdgeRequestEvent() {
```
[the remark this answers is not in the document]
Comment (kevinlind) [reply Tests/UnitTests/EdgeBridgeTests.swift:51]: What would be validated with the integration tests? As there is no Analytics response from Edge Network, we wouldn't be able to validate that the format we send it correct per Analytic's mapping.
The current functional tests validate the network request sent from the Edge extension. There are currently three tests 1) MobileCore.trackAction(), 2) MobileCore.trackState(), and 3) using Rules to dispatch a track call triggered from MobileCore.collectPii().

[the remark this answers is not in the document]
Comment (emdobrin) [reply Tests/UnitTests/EdgeBridgeTests.swift:51]: Cool, I was referring to similar tests but looks like we already have them covered. I agree, for upstream integration tests there is not much we can assert on, we can skip that.

File: Sources/EdgeBridgeConstants.swift [hunk 3]
```diff
@@ -25,4 +25,24 @@ enum EdgeBridgeConstants {
...
         static let EVENT_TYPE = "analytics.track"
     }
+
+    enum AnalyticsKeys {
+        static let ACTION = "action"
+        static let STATE = "state"
+        static let CONTEXT_DATA = "contextdata"
+    }
+
+    enum EdgeKeys {
```
Comment (emdobrin) [inline Sources/EdgeBridgeConstants.swift:35]: I got a little confused on the naming here since this is the Analytics format. renaming suggestion:
AnalyticsKeys -> MobileCoreKeys
EdgeKeys/Values -> AnalyticsKeys/Values

File: Sources/EdgeBridge.swift [hunk 4]
```diff
@@ -107,4 +107,76 @@ public class EdgeBridge: NSObject, Extension {
...
+                .map { (String($0.dropFirst(EdgeBridgeConstants.EdgeValues.PREFIX.count)), $1) }
+            analyticsData = Dictionary(uniqueKeysWithValues: prefixedDataArray) as? [String: Any] ?? [:]
+
+            let nonprefixedData = contextData.filter { !$0.key.hasPrefix(EdgeBridgeConstants.EdgeValues.PREFIX) }
+            if !nonprefixedData.isEmpty {
+                analyticsData[EdgeBridgeConstants.EdgeKeys.CONTEXT_DATA] = nonprefixedData
+            }
+        }
+
+        if let action = mutableData.removeValue(forKey: EdgeBridgeConstants.AnalyticsKeys.ACTION) as? String {
```
[the remark this answers is not in the document]
Comment (emdobrin) [reply Sources/EdgeBridge.swift:166]: In Analytics we set pageName=appid if none is populated, but I am not sure this is a requirement, especially considering that other Edge requests will have a missing action/pagename. @kevinlind can you follow-up on this?

(ref Analytics https://github.com/adobe/aepsdk-analytics-ios/pull/97)

[the remark this answers is not in the document]
Comment (kevinlind) [reply Sources/EdgeBridge.swift]: I can follow up on this next week when I'm back. 
Do you know where that referenced Analytics issue came from? Was it a customer request? As Edge Bridge only handles track API calls and not internal track events, I was just passing the `action` and `state` values as they are set by the customer. 
In direct to Analytics, it does appear the pageName is always set with a value, either the `state` value or the application ID if `action` is set or `state` is empty string. I'll follow up on this to see if this behavior is to be preserved.

File: Tests/UnitTests/EdgeBridgeTests.swift [hunk 5]
```diff
@@ -48,14 +48,248 @@ class EdgeBridgeTests: XCTestCase, AnyCodableAsserts {
...
+                          source: EventSource.requestContent,
+                          data: [
+                            "action": "action name",
+                            "key2": "value2",
+                            "contextdata": [
+                                "&&events": "event1,event2,event3,event4,event12,event13",
+                                "&&products": ";product1;1;5.99;event12=5.99;evar5=merchEvar5,;product2;2;10.99;event13=6;eVar6=mercheVar6",
+                                "&&c1": "propValue1",
+                                "&&cc": "USD",
+                                "key1": "value1"
```
Comment (emdobrin) [inline Tests/UnitTests/EdgeBridgeTests.swift:241]: @kevinlind same comment as for the Android version of this, can you add some negative testing null/empties, key name "&&", other special characters

File: Sources/EdgeBridge.swift [hunk 6]
```diff
@@ -91,20 +91,138 @@ public class EdgeBridge: NSObject, Extension {
...
-                                                   source: EventSource.requestContent,
-                                                   data: xdmEventData)
+            let event = parentEvent.createChainedEvent(name: EdgeBridgeConstants.EventNames.EDGE_BRIDGE_REQUEST,
+                                                       type: EventType.edge,
+                                                       source: EventSource.requestContent,
+                                                       data: xdmEventData)
 
-        runtime.dispatch(event: event)
+            runtime.dispatch(event: event)
+        } else {
```
[the remark this answers is not in the document]
Comment (emdobrin) [reply Sources/EdgeBridge.swift:110]: +1 on logging a message too

File: Tests/UnitTests/EdgeBridgeTests.swift [hunk 7]
```diff
@@ -67,24 +104,774 @@ class EdgeBridgeTests: XCTestCase, AnyCodableAsserts {
...
+                "eventType": "analytics.track"
+              }
+            }
+        """
+
+        assertEqual(expected: getAnyCodable(expectedJSON)!, actual: getAnyCodable(dispatchedEvent))
+    }
+
+    // Tests event is not dispatched is no track data is available
+    func testHandleTrackEvent_withNoMappedData_doesNotDispatchEvent() {
```
[the remark this answers is not in the document]
Comment (emdobrin) [reply Tests/UnitTests/EdgeBridgeTests.swift:484]: might be good to add a test for the cases where the track events were dispatched to event hub without using the track APIs.

File: Sources/EdgeBridge.swift [hunk 8]
```diff
@@ -91,20 +91,138 @@ public class EdgeBridge: NSObject, Extension {
...
+        var mutableData = data // mutable copy of data
+        var analyticsData: [String: Any] = [:] // __adobe.analytics data
+
+        if let contextData = mutableData.removeValue(forKey: EdgeBridgeConstants.MobileCoreKeys.CONTEXT_DATA) as? [String: Any], !contextData.isEmpty {
+            var prefixedData: [String: Any] = [:]
+            var nonprefixedData: [String: Any] = [:]
+
+            let cleanedContextData = cleanContextData(contextData)
+            for (key, value) in cleanedContextData {
+                if key.isEmpty {
```
[the remark this answers is not in the document]
Comment (emdobrin) [reply Sources/EdgeBridge.swift:168]: and maybe use debug level for these logs as they can get really chatty for an incorrect implementation

File: Sources/EdgeBridge.swift [hunk 9]
```diff
@@ -91,20 +91,138 @@ public class EdgeBridge: NSObject, Extension {
...
+    ///
+    ///  Note, empty keys are not allowed and ignored.
+    ///
+    /// - Parameter data: track event data
+    /// - Returns: data formatted for the Analytics Edge translator.
+    private func formatData(_ data: [String: Any]) -> [String: Any] {
+        var mutableData = data // mutable copy of data
+        var analyticsData: [String: Any] = [:] // __adobe.analytics data
+
+        if let contextData = mutableData.removeValue(forKey: EdgeBridgeConstants.MobileCoreKeys.CONTEXT_DATA) as? [String: Any], !contextData.isEmpty {
```
[the remark this answers is not in the document]
Comment (kevinlind) [reply Sources/EdgeBridge.swift:162]: Yes, a bit of an overkill but being more robust is better. I like it.

File: Tests/UnitTests/EdgeBridgeTests.swift [hunk 10]
```diff
@@ -67,24 +104,774 @@ class EdgeBridgeTests: XCTestCase, AnyCodableAsserts {
...
+                          source: EventSource.requestContent,
+                          data: [
+                            "contextdata": [
+                                "keyString": "valueString",
+                                "keyNumber": 5,
+                                "keyCharacter": char,
+                                "&&v1": "evar1",
+                                "&&v2": 10,
+                                "&&v3": char,
+                                "keyDict": ["hello": "world"],
```
[the remark this answers is not in the document]
Comment (kevinlind) [reply Tests/UnitTests/EdgeBridgeTests.swift:829]: I'll add that. It is possible to add additional data to the event through Rules, so it's good to test that the data is preserved.

File: Sources/EdgeBridge.swift [hunk 11]
```diff
@@ -107,4 +114,133 @@ public class EdgeBridge: NSObject, Extension {
...
+    ///
+    ///  Note, empty keys are not allowed and ignored.
+    ///
+    /// - Parameter data: track event data
+    /// - Returns: data formatted for the Analytics Edge translator.
+    private func formatData(_ data: [String: Any]) -> [String: Any] {
+        var mutableData = data // mutable copy of data
+        var analyticsData: [String: Any] = [:] // __adobe.analytics data
+
+        if let contextData = mutableData.removeValue(forKey: EdgeBridgeConstants.MobileCoreKeys.CONTEXT_DATA) as? [String?: Any?], !contextData.isEmpty {
```
Comment (emdobrin) [inline Sources/EdgeBridge.swift:164]: regarding `as? [String?: Any?]` are nil keys accepted in core event data?

File: Tests/UnitTests/EdgeBridgeTests.swift [hunk 12]
```diff
@@ -875,6 +909,59 @@ class EdgeBridgeTests: XCTestCase, AnyCodableAsserts {
...
+                "key": "value",
+                "&&c1": "prop1",
+                "keyEmptyValue": "",
+                "keyNilValue": nil,
+                "": "valueEmptyKey",
+                "additionalData": [
+                    "key": "value",
+                    nil: "valueNilKey"
+                ]
+            ],
```
Comment (emdobrin) [inline Tests/UnitTests/EdgeBridgeTests.swift:1048]: nit: you may declare this map once and reuse it for the expectedData

**Review by emdobrin — APPROVED** [review]
